File: cli/internal/connectapi/link_previews.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
	corev1 "hmans.de/chatto/internal/pb/chatto/core/v1"
)

func (s *messageService) FetchLinkPreview(ctx context.Context, req *connect.Request[apiv1.FetchLinkPreviewRequest]) (*connect.Response[apiv1.FetchLinkPreviewResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	preview, attachment, err := s.api.core.GetLinkPreviewForComposer(ctx, caller.UserID, req.Msg.GetRoomId(), req.Msg.Url)
	if err != nil {
		return nil, connectError(err)
	}
	if attachment != nil {
		asset := (&attachmentMapper{api: s.api}).asset(attachment, caller.UserID, assetThumbnailOptions(nil))
		previewURL := ""
		if asset.GetThumbnailAssetUrl() != nil {
			previewURL = asset.GetThumbnailAssetUrl().GetUrl()
		} else if asset.GetAssetUrl() != nil {
			previewURL = asset.GetAssetUrl().GetUrl()
		}
		return connect.NewResponse(&apiv1.FetchLinkPreviewResponse{
			ImportedAttachment: &apiv1.ImportedLinkAttachment{
				AssetId:     asset.GetId(),
				Filename:    asset.GetFilename(),
				ContentType: asset.GetContentType(),
				Size:        asset.GetSize(),
				Width:       asset.GetWidth(),
				Height:      asset.GetHeight(),
				PreviewUrl:  previewURL,
			},
		}), nil
	}
	if preview == nil {
		return connect.NewResponse(&apiv1.FetchLinkPreviewResponse{}), nil
	}
	tokenURL := preview.GetUrl()
	if tokenURL == "" {
		tokenURL = req.Msg.Url
	}
	token, err := s.api.core.CreateLinkPreviewToken(ctx, tokenURL)
	if err != nil {
		return nil, connectError(err)
	}

	return connect.NewResponse(&apiv1.FetchLinkPreviewResponse{
		Preview:      apiLinkPreview(s.api, preview),
		PreviewToken: token,
	}), nil
}

func apiLinkPreview(api *API, preview *corev1.LinkPreview) *apiv1.LinkPreview {
	if preview == nil {
		return nil
	}

	imageAssetID := preview.GetImageAssetId()
	if image := preview.GetImageAsset(); image != nil && image.GetId() != "" {
		imageAssetID = image.GetId()
	}

	imageURL := ""
	if imageAssetID != "" {
		imageURL = api.core.GetTransformedServerAssetURL(imageAssetID, 600, 314, "contain")
	}

	out := &apiv1.LinkPreview{
		Url: preview.GetUrl(),
	}
	if title := preview.GetTitle(); title != "" {
		out.Title = stringPtr(title)
	}
	if description := preview.GetDescription(); description != "" {
		out.Description = stringPtr(description)
	}
	if imageURL != "" {
		out.ImageUrl = stringPtr(imageURL)
	}
	if imageAssetID != "" {
		out.ImageAssetId = stringPtr(imageAssetID)
	}
	if siteName := preview.GetSiteName(); siteName != "" {
		out.SiteName = stringPtr(siteName)
	}
	if embedType := preview.GetEmbedType(); embedType != "" {
		out.EmbedType = stringPtr(embedType)
	}
	if embedID := preview.GetEmbedId(); embedID != "" {
		out.EmbedId = stringPtr(embedID)
	}
	return out
}

File: apps/frontend/src/lib/api-client/linkPreviews.ts
import { authHeaders, createChattoClient, handleAuthError } from './connect.js';
import { MessageService } from '@chatto/api-types/api/v1/messages_connect';
import type { LinkPreview } from '@chatto/api-types/api/v1/link_previews_pb';
import type { ImportedLinkAttachment } from '@chatto/api-types/api/v1/link_previews_pb';

export type LinkPreviewAPIConfig = {
  serverId?: string;
  baseUrl: string;
  bearerToken: string | null;
  onAuthenticationRequired?: (serverId: string) => void;
};

export type ComposerLinkPreview = {
  url: string;
  previewToken: string;
  title: string | null;
  description: string | null;
  imageUrl: string | null;
  imageAssetId: string | null;
  siteName: string | null;
  embedType: string | null;
  embedId: string | null;
};

export type ComposerImportedAttachment = {
  assetId: string;
  filename: string;
  contentType: string;
  size: bigint;
  width: number;
  height: number;
  previewUrl: string;
};

export type ComposerLinkResult =
  | { kind: 'preview'; preview: ComposerLinkPreview }
  | { kind: 'attachment'; attachment: ComposerImportedAttachment };

export function createLinkPreviewAPI(config: LinkPreviewAPIConfig) {
  const client = createChattoClient(MessageService, config);
  const headers = () => authHeaders(config);
  return {
    async fetchLinkPreview(url: string, roomId?: string): Promise<ComposerLinkResult | null> {
      try {
        const response = await client.fetchLinkPreview({ url, roomId }, { headers: headers() });
        const attachment = composerImportedAttachment(response.importedAttachment);
        if (attachment) return { kind: 'attachment', attachment };
        const preview = composerLinkPreview(response.preview, response.previewToken);
        return preview ? { kind: 'preview', preview } : null;
      } catch (err) {
        return handleAuthError(config, err);
      }
    }
  };
}

function composerLinkPreview(
  preview: LinkPreview | undefined,
  previewToken: string
): ComposerLinkPreview | null {
  if (!preview || !previewToken) return null;
  return {
    url: preview.url,
    previewToken,
    title: preview.title || null,
    description: preview.description || null,
    imageUrl: preview.imageUrl || null,
    imageAssetId: preview.imageAssetId || null,
    siteName: preview.siteName || null,
    embedType: preview.embedType || null,
    embedId: preview.embedId || null
  };
}

function composerImportedAttachment(
  asset: ImportedLinkAttachment | undefined
): ComposerImportedAttachment | null {
  if (!asset?.assetId || !asset.previewUrl) return null;
  return {
    assetId: asset.assetId,
    filename: asset.filename,
    contentType: asset.contentType,
    size: asset.size,
    width: asset.width,
    height: asset.height,
    previewUrl: asset.previewUrl
  };
}
